### PR TITLE
GeNN 4 updates

### PR DESCRIPTION
**Pull request by neworderofjamie**
Mostly very minor syntactic changes but also some new functionality and efficiency:
- GeNN 4 lets you copy variables to/from GPU individually so monitors now use this functionality
- PyGeNN now provides an API for accessing sparse indices

Could you check your models still work and that I haven't done anything dumb in this!

### CODE REVIEW

**Comment by chanokin** [comment]
I reinstalled GeNN from master and pynn_genn from the genn_4 branch. If I run non-plastic it seems to work ok (limited testing) but if I activate STDP there are some errors during compilation (CPU based):

```
synapseUpdate.cc:19:229: error: ‘max’ was not declared in this scope
```
```
synapseUpdate.cc:53:244: error: ‘min’ was not declared in this scope
```
```
swig/python detected a memory leak of type 'CodeGenerator::SingleThreadedCPU::Backend *', no destructor found.
```

**Comment by neworderofjamie** [comment]
Ahh shit - good spot! It **should** only effect CPU-only mode but will fix ASAP

File: README.MD [hunk 1]
```diff
@@ -2,17 +2,10 @@
...
- - On Mac OS X only, set your newly created library's name with ``install_name_tool -id "@loader_path/libgenn_DYNAMIC.dylib" pygenn/genn_wrapper/libgenn_DYNAMIC.dylib`` (you will need to replace ``libgenn_DYNAMIC`` with ``libgenn_CPU_ONLY_DYNAMIC`` if you do not have an NVIDIA GPU)
- - Install with setup tools using ``python setup.py develop`` command
- 
-### Installing the PyNN backend for GeNN
- - Clone this project
- - Install with setuptools using ``python setup.py develop`` command
+## Installation
+ 1. Follow the instructions in https://github.com/genn-team/genn/blob/master/pygenn/README.md to install PyGeNN.
+ 2. Clone this project
+ 3. Install with setuptools using ``python setup.py develop`` command
```
Comment (chanokin) [inline README.MD]: This still seems to need GeNN to have LIBRARY_DIRECTORY pointing at pygenn/genn_wrapper

File: pynn_genn/__init__.py [hunk 2]
```diff
@@ -46,7 +46,10 @@ def setup(timestep=DEFAULT_TIMESTEP, min_delay=DEFAULT_MIN_DELAY,
     simulator.state.max_delay = max_delay
     simulator.state.mpi_rank = extra_params.get("rank", 0)
     simulator.state.num_processes = extra_params.get("num_processes", 1)
-    simulator.state.model.use_cpu = extra_params.get("use_cpu", None)
+
+    # If backend is specified, use that
+    if "backend" in extra_params:
+        simulator.state.model.use_backend = extra_params["backend"]
```
Comment (chanokin) [inline pynn_genn/__init__.py]: Fixed the CPU-based choice issue

**Pull request merged**